From a0f2605263d61a6bbf999d1250525d2b5ef27222 Mon Sep 17 00:00:00 2001
From: Stanley Chu <yschu@nuvoton.com>
Date: Thu, 27 Jun 2024 16:28:17 +0800
Subject: [PATCH] i3c: send 7e before private read

Signed-off-by: Stanley Chu <yschu@nuvoton.com>
---
 drivers/i3c/master/npcm845-i3c-master.c | 64 +++++++++++++++++++++----
 1 file changed, 54 insertions(+), 10 deletions(-)

diff --git a/drivers/i3c/master/npcm845-i3c-master.c b/drivers/i3c/master/npcm845-i3c-master.c
index 8a7fa1c0983c..3fab9a063e1e 100644
--- a/drivers/i3c/master/npcm845-i3c-master.c
+++ b/drivers/i3c/master/npcm845-i3c-master.c
@@ -1350,17 +1350,48 @@ static int npcm_i3c_master_wait_for_complete(struct npcm_i3c_master *master)
 	return count;
 }
 
+static int npcm_i3c_send_broadcast(struct npcm_i3c_master *master)
+{
+	u32 reg;
+	int ret;
+
+	writel(NPCM_I3C_MCTRL_REQUEST_START_ADDR |
+	       NPCM_I3C_MCTRL_TYPE_I3C |
+	       NPCM_I3C_MCTRL_IBIRESP_AUTO |
+	       NPCM_I3C_MCTRL_DIR(0) |
+	       NPCM_I3C_MCTRL_ADDR(I3C_BROADCAST_ADDR) |
+	       NPCM_I3C_MCTRL_RDTERM(NPCM_I3C_MAX_IBI_PAYLOAD_SIZE),
+	       master->regs + NPCM_I3C_MCTRL);
+	ret = readl_poll_timeout(master->regs + NPCM_I3C_MSTATUS, reg,
+				 NPCM_I3C_MSTATUS_MCTRLDONE(reg), 0, 1000);
+	if (ret)
+		return -EIO;
+
+	if (NPCM_I3C_MSTATUS_IBIWON(reg)) {
+		ret = npcm_i3c_master_handle_ibiwon(master, true);
+		if (ret) {
+			dev_err(master->dev, "xfer read: handle ibi event fail, ret=%d\n", ret);
+			return -EIO;
+		}
+
+		/* Clear COMPLETE status of this IBI transaction */
+		writel(NPCM_I3C_MINT_COMPLETE, master->regs + NPCM_I3C_MSTATUS);
+		return 1;
+	}
+
+	return 0;
+}
+
 static int npcm_i3c_master_xfer(struct npcm_i3c_master *master,
 			       bool rnw, unsigned int xfer_type, u8 addr,
 			       u8 *in, const u8 *out, unsigned int xfer_len,
 			       unsigned int *read_len, bool continued,
-			       bool use_dma)
+			       bool use_dma, bool first)
 {
 	u32 reg, rdterm = *read_len, mstatus;
 	int ret, i, count, space;
 	unsigned long flags;
 	unsigned long start;
-	u32 ibiresp;
 
 	if (rdterm > NPCM_I3C_MAX_RDTERM)
 		rdterm = NPCM_I3C_MAX_RDTERM;
@@ -1417,15 +1448,28 @@ static int npcm_i3c_master_xfer(struct npcm_i3c_master *master,
 	}
 
 	start = jiffies;
-	/*
-	 * IBI payload size may be larger than rdterm, use manual IBI response
-	 * for read operation to set the proper RDTERM value in IBI ack request.
-	 */
-	ibiresp = rnw ? NPCM_I3C_MCTRL_IBIRESP_MANUAL : NPCM_I3C_MCTRL_IBIRESP_AUTO;
+
 retry_start:
+	if (first && rnw) {
+		/* Send 7E first to avoid collision during master read */
+		ret = npcm_i3c_send_broadcast(master);
+		if (ret < 0) {
+			dev_info(master->dev, "send 7e error\n");
+			goto unlock_exit;
+		}
+		if (time_after(jiffies, start + msecs_to_jiffies(1000))) {
+			dev_info(master->dev, "abnormal ibiwon events\n");
+			goto unlock_exit;
+		}
+		if (ret == 1)
+			goto retry_start;
+	}
+	if (use_dma)
+		npcm_i3c_master_start_dma(master);
+
 	writel(NPCM_I3C_MCTRL_REQUEST_START_ADDR |
 	       xfer_type |
-	       ibiresp |
+	       NPCM_I3C_MCTRL_IBIRESP_AUTO |
 	       NPCM_I3C_MCTRL_DIR(rnw) |
 	       NPCM_I3C_MCTRL_ADDR(addr) |
 	       NPCM_I3C_MCTRL_RDTERM(rdterm),
@@ -1441,7 +1485,7 @@ static int npcm_i3c_master_xfer(struct npcm_i3c_master *master,
 		/* Stop RX DMA to prevent it from receving the ibi payload */
 		if (use_dma && rnw)
 			npcm_i3c_master_stop_dma(master);
-		ret = npcm_i3c_master_handle_ibiwon(master, !rnw);
+		ret = npcm_i3c_master_handle_ibiwon(master, true);
 		if (ret) {
 			dev_err(master->dev, "xfer read: handle ibi event fail, ret=%d\n", ret);
 			goto unlock_exit;
@@ -1577,7 +1621,7 @@ static void npcm_i3c_master_start_xfer_locked(struct npcm_i3c_master *master)
 		ret = npcm_i3c_master_xfer(master, cmd->rnw, xfer->type,
 					  cmd->addr, cmd->in, cmd->out,
 					  cmd->len, &cmd->read_len,
-					  cmd->continued, cmd->use_dma);
+					  cmd->continued, cmd->use_dma, (i == 0));
 		if (ret)
 			break;
 	}
-- 
2.34.1

